rfc/20240824-provider-cache-locking.md

```diff
@@ -0,0 +1,69 @@
+
+
+Mixture of missing platform/corrupt lock files with valid lock files.
+
+The provider lock file may have been generated on a system with a different architecture or may have become corrupt due to a variety of reasons. When this happens, the global cache may not match the lock file and force a re-download of the provider. This will cause potentially unexpected downloads in the above scenarios.
+
+### Technical Approach
+
+This proposal hinges on a stable and consistent cross-platform lock. The codebase already contains this in the form of locking local state files.  This code is battle tested and overall quite simple to use.  With some light refactoring, this filesystem locking code can be moved into it's own internal package and used both to lock providers and to lock the local state file.
+
+The existing file locking should be safe on any local filesystems, but should be used with caution on shared volumes such as legacy NFS shares who do not provide strong locking consistency. We will add explicit warnings to the documentation which recommend against using networked filesystems for this use case.
+
+Within the provider installation code, the whole section which inspects and links to the current providers available should be locked at the provider level.  The lock should be done at this granularity specifically for the complicating factors mentioned above.
+
+Additionally, we should make the package installer smarter and able to check the files in the cache against the downloaded version. This prevents a bad cache entry from overwriting a valid entry which may already been in use by another process.
+
+### Open Questions
+
+* Is a filesystem lock syscall safe enough? It is industry standard and battle tested.  Are there any additional caveats that should be mentioned in the docs?
+  - Yes. As mentioned above, we should document that networked filesystems are not recommended for this use case.
+
+### Future Considerations
+
+A considerable amount of time is spent scanning the provider cache.  Instead, it should probably be refactored to only read the provider metadata necessary at any given time.  This will offer some significant performance bonuses on large configurations.
+
+## Potential Alternatives
+
+Terragrunt/Gruntwork provides a http provider mirror that can be run locally and maintains a provider cache (archives) external to OpenTofu itself. While this does have some advantages, it still requires each provider to be extracted into the local provider cache folder, taking additional time and space.  It may be a safer alternative to using a networked filesystem between systems.
```

Comment:
FWIW - It is not that great. Terrwgraunt starts local https proxy server within single invocation of `terragrunt run` / `terragrunt run-all` and stops the server on the end of the command (there is a ticket to improve this: https://github.com/gruntwork-io/terragrunt/issues/3363).

So currently, when you run in parallel multiple `terragrunt run` there is still racing. That's the case for atlantis+terragrunt setup for example. So I would not consider terragrunt as a "potential alternative". We do need this fixed on opentofu level.

Reply:
Being honest, I'm not sure I'll have time to dig back into this until the new year.  I'll try though.